### DISCUSSION TITLE
Kit generate CR line breaks when writing to files

### DISCUSSION THREAD

**Discussion by pomdtr** [johnlindquist/kit#375] [Errors]
When saving the content of the clipboard or the current selection to a file, I get windows-style line breaks:

```javascript
const clipboardContent = await paste();
const selectedText = await kit.getSelectedText();
const tempfile = `${kit.tempdir()}/input.txt`

if (selectedText) {
  await kit.writeFile(tempfile, selectedText, {flag: 'w', encoding: 'utf-8'})
} else if (clipboardContent) {
  await kit.writeFile(tempfile, clipboardContent,  {flag: 'w', encoding: 'utf-8'})
}
```

Is there a way to stop this behavior ?

**Comment by pomdtr** [comment]
In order to obtain the expected linebreaks, I need to use the command `tr '\r' '\n'`.

Here is the full script:

```javascript
// Menu: Pipe to Command
// Decription: Manipulate the selected text or clipboard content using the shell
// Author: pomdtr

process.env.PATH = `${process.env.HOME}/.local/bin:/usr/local/bin/:${process.env.PATH}`;
var os = import('os'); os.EOL

function codeblock(text) {
  const triple_backquote = "```";
  return `${triple_backquote}shell\n${text}\n${triple_backquote}`;
}

let command = "";

const clipboardContent = await paste();
const selectedText = await kit.getSelectedText();
const tempfile = `${kit.tempdir()}/input.txt`

if (selectedText) {
  await kit.writeFile(tempfile, selectedText, {flag: 'w', encoding: 'utf-8'})
} else if (clipboardContent) {
  await kit.writeFile(tempfile, clipboardContent,  {flag: 'w', encoding: 'utf-8'})
}

let [stdout, stderr, code] = [null, null, 0];
let panelContent = await kit.readFile(tempfile, {encoding: "utf-8"})

while (true) {
  command = await arg(
    {
      placeholder: "Command:",
      className: "px-2",
      input: command,
    },
    md(codeblock(panelContent))
  );

  if (command) {
    ({stdout, stderr, code} = exec(`tr '\r' '\n' < ${tempfile} | ${command}`))
    panelContent = code == 0 ? stdout : stderr;
  } else {
    panelContent = await kit.readFile(tempfile, {encoding: "utf-8"});
  }

  await copy(panelContent);
}
```

**Comment by johnlindquist** [comment]
Thanks for finding this. I dug into what was going on, here's what I found:

## Problem

The internal `getSelectedText` has this line:
```js
  let selectedText = await applescript(
    String.raw`get the clipboard`
  )
```

The `applescript` helper is simply a wrapper around a bash command:

```bash
  let { stdout, stderr } = exec(
    `osascript -e '${formattedScript}'`,
    options
  )
```

So at its core, `getSelectedText` is really just this bash command (you can run this in your terminal):

```bash
osascript -e 'get the clipboard'
```

So if we actually create a file and check for line endings:

```bash
osascript -e 'get the clipboard' > ~/.kenv/tmp/testing-line-breaks/test.txt
file ~/.kenv/tmp/testing-line-breaks/test.txt
```
The result is:
`tmp/testing-line-breaks/test.txt: ASCII text, with CRLF, CR line terminators`

We don't want `CRLF` 😬

## Quick Solution

"Don't use AppleScript" 😇. I'll just replace:

this:
```js
  let selectedText = await applescript(
    String.raw`get the clipboard`
  )
```

with this:
```js
let selectedText = await paste() // "paste" is from npm's "clipboardy"
```

and the problem goes away.

## Long-term solution

Later this year, I'll be removing all the AppleScript code in favor of Rust w/ osx bindings and I'll make sure any returned strings have proper line-endings. The plan is to bring Script Kit to Windows and Rust cross-platform bindings should fit the purpose nicely. AppleScript is _slow_ when dealing many OS-level apis (window bounds, displays, running apps, etc) and I'm excited to build much more robust OS-level tools (and keep them cross-platform) with Rust.
